sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamMetrics.scala

```diff
@@ -0,0 +1,252 @@
+    sourceTriggerInfo(source).put(key, value.toString)
+  }
+
+  def reportNumRows(inputRows: Map[Source, Long], outputRows: Option[Long]): Unit = synchronized {
+    numInputRows ++= inputRows
+    numOutputRows = outputRows
+  }
+
+  def reportTriggerFinished(): Unit = synchronized {
+    require(currentTriggerStartTimestamp >= 0)
+    val currentTriggerFinishTimestamp = triggerClock.getTimeMillis()
+    reportTriggerInfo(FINISH_TIMESTAMP, currentTriggerFinishTimestamp)
+    reportTriggerInfo(STATUS_MESSAGE, "")
+    reportTriggerInfo(ACTIVE, false)
+
+    // Report number of rows
+    val totalNumInputRows = numInputRows.values.sum
+    reportTriggerInfo(NUM_INPUT_ROWS, totalNumInputRows)
+    reportTriggerInfo(NUM_OUTPUT_ROWS, numOutputRows.getOrElse(0))
+    numInputRows.foreach { case (s, r) =>
+      reportSourceTriggerInfo(s, NUM_SOURCE_INPUT_ROWS, r)
+    }
+
+    val currentTriggerDuration = currentTriggerFinishTimestamp - currentTriggerStartTimestamp
+    val previousInputIntervalOption = if (previousTriggerStartTimestamp >= 0) {
+      Some(currentTriggerStartTimestamp - previousTriggerStartTimestamp)
+    } else None
+
+    // Update input rate = num rows received by each source during the previous trigger interval
+    // Interval is measures as interval between start times of previous and current trigger.
+    //
+    // TODO: Instead of trigger start, we should use time when getOffset was called on each source
+    // as this may be different for each source if there are many sources in the query plan
+    // and getOffset is called serially on them.
+    if (previousInputIntervalOption.nonEmpty) {
+      sources.foreach { s =>
+        inputRates(s).update(numInputRows.getOrElse(s, 0), previousInputIntervalOption.get)
+      }
+    }
+
+    // Update processing rate = num rows processed for each source in current trigger duration
+    sources.foreach { s =>
+      processingRates(s).update(numInputRows.getOrElse(s, 0), currentTriggerDuration)
+    }
+
+    // Update output rate = num rows output to the sink in current trigger duration
+    outputRate.update(numOutputRows.getOrElse(0), currentTriggerDuration)
+    logDebug("Output rate updated to " + outputRate.currentRate)
+
+    // Update latency = if data present, 0.5 * previous trigger interval + current trigger duration
+    if (previousInputIntervalOption.nonEmpty && totalNumInputRows > 0) {
+      latency = Some((previousInputIntervalOption.get.toDouble / 2) + currentTriggerDuration)
+    } else {
+      latency = None
+    }
+
+    previousTriggerStartTimestamp = currentTriggerStartTimestamp
+    currentTriggerStartTimestamp = -1
+  }
+
+  // =========== Getter methods ===========
+
+  def currentInputRate(): Double = synchronized {
+    // Since we are calculating source input rates using the same time interval for all sources
+    // it is fine to calculate total input rate as the sum of per source input rate.
+    inputRates.map(_._2.currentRate).sum
```

Comment:
For the record, resolved. Synchronization has been simplified by the stream execution thread updating a volatile reference to the query status, every time any of components used in the status is updated. StreamingQuery.status only read the current value of the volatile reference. See `updateStatus()` in the PR.